server/routes.go

```diff
@@ -1549,42 +1568,51 @@ func (s *Server) ChatHandler(c *gin.Context) {
 			}
 
 			if r.Done {
+				if sb.Len() > 0 {
+					res.Message.Content = sb.String()
+				}
 				res.DoneReason = r.DoneReason.String()
 				res.TotalDuration = time.Since(checkpointStart)
 				res.LoadDuration = checkpointLoaded.Sub(checkpointStart)
 			}
 
-			// TODO: tool call checking and filtering should be moved outside of this callback once streaming
-			// however this was a simple change for now without reworking streaming logic of this (and other)
-			// handlers
-			if req.Stream != nil && !*req.Stream || len(req.Tools) == 0 {
-				ch <- res
-				return
-			}
-
-			// Streaming tool calls:
-			// If tools are recognized, use a flag to track the sending of a tool downstream
-			// This ensures that content is cleared from the message on the last chunk sent
 			sb.WriteString(r.Content)
-			if toolCalls, ok := m.parseToolCalls(sb.String()); ok {
-				res.Message.ToolCalls = toolCalls
-				for i := range toolCalls {
-					toolCalls[i].Function.Index = toolCallIndex
-					toolCallIndex++
+			if len(req.Tools) > 0 && checkToolCall {
+				slog.Debug("parse tool calls", "content", sb.String(), "templateToolToken", templateToolToken)
+				toolCalls, partial, err := ParseToolCalls(sb.String(), templateToolToken, tmpl)
+				if err == nil {
+					if partial {
+						// circuit break to remove tool end token
+						if len(toolCalls) > 0 {
+							sb.Reset()
+						}
+						// If the tool call is partial, we need to wait for the next chunk
+						return
+					}
+					res.Message.ToolCalls = toolCalls
+					for i := range toolCalls {
+						toolCalls[i].Function.Index = toolCallIndex
+						toolCallIndex++
+					}
+					res.Message.Content = ""
+					ch <- res
+					// Only way to have multiple calls is to have [] which is derived or provided
+					// This case occurs when the tool call is a json block - do not allow tool calls again
```

Comment:
Not totally clear what this means - ideally ParseToolCalls above handles this case